net/redisclient.go

```diff
@@ -87,6 +101,96 @@ const (
 	defaultConnMetricsInterval = 60 * time.Second
 )
 
+// JumpHash
+// https://arxiv.org/pdf/1406.2294.pdf
+type JumpHash struct {
+	hash   hash.Hash64
+	shards []string
+}
+
+func NewJumpHash(shards []string) redis.ConsistentHash {
+	return &JumpHash{
+		hash:   fnv.New64(),
+		shards: shards,
+	}
+}
+
+func (j *JumpHash) Get(k string) string {
+	_, err := j.hash.Write([]byte(k))
+	if err != nil {
+		log.Fatalf("Failed to write %s to hash: %v", k, err)
+	}
+
+	key := j.hash.Sum64()
+	h := jump.Hash(key, len(j.shards))
+	return j.shards[int(h)%len(j.shards)]
```

Comment:
https://github.com/dgryski/go-jump/blob/e1f439676b570800a863c37b78d1f868f51a85fc/jump.go#L10
> Hash consistently chooses a hash bucket number in the range [0, numBuckets) for the given key. numBuckets must be >= 1

so there is no need to use modulo operation.

But I think we have another problem here. 
Jump hash (as far as I can tell) is designed for the application where buckets/nodes are added sequentially and never removed from the middle.
When redis shard goes down, redis client recreates the hash with a list of available shards and thus breaks jump hash requirement of not removing the shard.

> This ability to add or remove buckets in any order can be valuable for cache servers where the servers are purely a performance improvement. But for data storage applications, where each bucket represents a different shard of the data, it is not acceptable for shards to simply disappear, because that shard is only place where the corresponding data is stored. Typically this is handled by either making the shards
redundant (with several replicas), or being able to quickly recover a replacement, or accepting lower availability for some data. Server death thus does not cause reallocation of data; only capacity changes do. This means that shards can be assigned numerical ids in increasing order as they are added, so that the active bucket ids always fill the range [0, num_buckets).

I.e. jump seems not applicable for our usecase because it will not provide the properties it have when conditions are not statisfied.

Reply:
There is never a condition that will remove shards from the ring in the middle permanently. It would be an incident if it would happen because of failing instance, so I think it's fine. I think we discussed this already in person. :)

Reply:
> in the middle permanently

Ok, but shard may go down for some time and that would shift key distribution.

Also jump hash assumes node list keeps the order on growth (i.e. existing shard keeps its index in the list). Redis client hash does not guarantee that shard list would contain newly added shard at the end and preserve the order.
Maybe it worth adding all assumptions as a comment.

The modulo is still redundant and could be simplified to `return j.shards[int(h)]`

Overall the only benefit of this algo over say rendevouz is that it "fast" but I think the performance gain is negligible compared to string hash calculation costs given there are expected only a handful of nodes.

Reply:
deleted the modulo operation